extensions/ql-vscode/src/model-editor/methods-usage/methods-usage-data-provider.ts

```diff
@@ -117,23 +116,24 @@ export class MethodsUsageDataProvider
     if (isExternalApiUsage(item)) {
       return undefined;
     } else {
-      return this.externalApiUsages.find((e) => e.usages.includes(item));
+      return item.method;
     }
   }
 
-  public resolveCanonicalUsage(usage: Usage): Usage | undefined {
+  public resolveCanonicalUsage(usage: Usage): UsageWithMethod | undefined {
     for (const externalApiUsage of this.externalApiUsages) {
       for (const u of externalApiUsage.usages) {
         if (usagesAreEqual(u, usage)) {
-          return u;
+          return { ...u, method: externalApiUsage };
         }
       }
     }
     return undefined;
   }
 }
 
-export type MethodsUsageTreeViewItem = ExternalApiUsage | Usage;
+type UsageWithMethod = Usage & { method: ExternalApiUsage };
+export type MethodsUsageTreeViewItem = ExternalApiUsage | UsageWithMethod;
```

Comment:
I agree and I like this approach. Going a bit further and introducing types like you included here would make things a bit cleaner. At least at first it was attractive to reuse the existing types as they are instead of making new ones, but it just causes more problems because then you can't make changes or include new data. We should just make the switch to more custom types.

Reply:
Ok, upon trying this out, there's a problem. The problem is that features in the tree view like revealing work on identifying exact tree items by their object identity. For VS Code to be able to reveal a particular item in the tree, we need to be able to pass to the `reveal` method the exact same object that we returned from `getChildren`.

This worked before because we weren't creating any new object, but we were just returning the same `ExternalApiUsage` and `Usage` objects we were given. But if we make new types then we will have to construct new objects, or essentially build the whole tree out of new objects which means redefining and duplicating the definition of `ExternalApiMethod`.

I suppose we could have it be defined as
```
interface MethodTreeViewItem {
  kind: "method";
  method: ExternalApiUsage;
  usages: UsageTreeViewItem[];
}

interface UsageTreeViewItem {
  kind: "usage";
  parent: MethodTreeViewItem;
  usage: Usage;
}
```
but this means the data is only semi-duplicated.

It also leads to problems in the `setState` function where we decide whether to re-render or not, because the state we're storing (i.e. `MethodTreeViewItem[]`) is not the same as the arguments we're given (i.e. `ExternalApiUsage[]`).

I think that my current implementation in this PR should also have this same problem because I'm creating a new object in `resolveCanonicalUsage` instead of returning an existing object. Somehow it does appear to be working though, and I'm not sure why. I have a few hypotheses:
- I might be wrong and it's fine to return a new object here.
- It might be fine because `getParent` still resolves to the correct canonical `ExternalApiUsage` object, and then it can't find the child so it highlights the first child which happens to be what we wanted anyway.
- It might be the javascript engine reusing memory for objects because it recognises that they're the same and we never mutate the object. Is this something javascript does? 🤷🏼 

Overall, we might be able to get it to work, but I'm not totally sure.

Reply:
Sorry I don't have a huge amount of time to think about this, what do you think about creating a tech debt issue?

Reply:
I think an issue would be more sensible than trying to think it through on a PR.